File: Faculdade/interpreter/teste.java
// Interface Expression (Expressão Abstrata)
interface Expression {
    int interpret();
}

// Classe TerminalExpression (Número)
class Number implements Expression {
    private int value;

    public Number(int value) {
        this.value = value;
    }

    @Override
    public int interpret() {
        return value;
    }
}

// Classe NonTerminalExpression (Soma)
class Addition implements Expression {
    private Expression leftExpression;
    private Expression rightExpression;

    public Addition(Expression leftExpression, Expression rightExpression) {
        this.leftExpression = leftExpression;
        this.rightExpression = rightExpression;
    }

    @Override
    public int interpret() {
        return leftExpression.interpret() + rightExpression.interpret();
    }
}

// Classe NonTerminalExpression (Subtração)
class Subtraction implements Expression {
    private Expression leftExpression;
    private Expression rightExpression;

    public Subtraction(Expression leftExpression, Expression rightExpression) {
        this.leftExpression = leftExpression;
        this.rightExpression = rightExpression;
    }

    @Override
    public int interpret() {
        return leftExpression.interpret() - rightExpression.interpret();
    }
}

// Classe Principal (Cliente)
public class InterpreterExample {
    public static void main(String[] args) {
        // Criando a expressão: (10 + 5) - 3
        Expression expression = new Subtraction(
            new Addition(new Number(10), new Number(5)),
            new Number(3)
        );

        // Interpretar e exibir o resultado
        int result = expression.interpret();
        System.out.println("Resultado da expressão: " + result); // Saída: 12
    }
}

File: Faculdade/interpreter/teste.py
from abc import ABC, abstractmethod

# Classe abstrata para expressões
class Expressao(ABC):
    @abstractmethod
    def interpretar(self):
        pass

# Expressão terminal para números
class Numero(Expressao):
    def __init__(self, valor):
        self.valor = valor

    def interpretar(self):
        return self.valor

# Expressão não terminal para soma
class Soma(Expressao):
    def __init__(self, esquerda, direita):
        self.esquerda = esquerda
        self.direita = direita

    def interpretar(self):
        return self.esquerda.interpretar() + self.direita.interpretar()

# Expressão não terminal para subtração
class Subtracao(Expressao):
    def __init__(self, esquerda, direita):
        self.esquerda = esquerda
        self.direita = direita

    def interpretar(self):
        return self.esquerda.interpretar() - self.direita.interpretar()

# Cliente
if __name__ == "__main__":
    # Representação da expressão: (10 + 5) - 3
    expressao = Subtracao(
        Soma(Numero(10), Numero(5)),
        Numero(3)
    )

    # Interpretar e exibir o resultado
    resultado = expressao.interpretar()
    print("Resultado da expressão:", resultado)  # Saída: Resultado da expressão: 12



#            Subtracao
#           /          \
#       Soma            Numero(3)
#      /    \
# Numero(10) Numero(5)
